ports/nrf/common-hal/audiopwmio/PWMAudioOut.c

```diff
@@ -0,0 +1,315 @@
+    }
+    self->sample = sample;
+    self->loop = loop;
+
+    uint32_t sample_rate = audiosample_sample_rate(sample);
+    uint32_t max_sample_rate = 62500;
+    if (sample_rate > max_sample_rate) {
+        mp_raise_ValueError_varg(translate("Sample rate too high. It must be less than %d"), max_sample_rate);
+    }
+    self->bytes_per_sample = audiosample_bits_per_sample(sample) / 8;
+
+    uint32_t max_buffer_length;
+    audiosample_get_buffer_structure(sample, /* single channel */ false,
+        &self->single_buffer, &self->signed_to_unsigned, &max_buffer_length,
+        &self->spacing);
+    if(max_buffer_length > UINT16_MAX) {
+        mp_raise_ValueError_varg(translate("Buffer length %d too big. It must be less than %d"), max_buffer_length, UINT16_MAX);
+    }
+    self->buffer_length = (uint16_t)max_buffer_length;
+    self->buffers[0] = m_malloc(self->buffer_length * 2 * sizeof(uint16_t), false);
+    if(!self->single_buffer)
+        self->buffers[1] = m_malloc(self->buffer_length * 2 * sizeof(uint16_t), false);
+
+
+    uint32_t top;
+    self->pwm->SEQ[0].REFRESH = self->pwm->SEQ[1].REFRESH = calculate_pwm_parameters(sample_rate, &top);
+    self->scale = top-1;
+    self->pwm->COUNTERTOP = top;
+
+    self->pwm->LOOP = 1;
+    audiosample_reset_buffer(self->sample, false, 0);
+    activate_audiopwmout_obj(self);
+    fill_buffers(self, 0);
+    self->pwm->SEQ[1].PTR = self->pwm->SEQ[0].PTR;
+    self->pwm->SEQ[1].CNT = self->pwm->SEQ[0].CNT;
+    self->pwm->EVENTS_SEQSTARTED[0] = 0;
+    self->pwm->EVENTS_SEQSTARTED[1] = 0;
+    self->pwm->EVENTS_STOPPED = 0;
+    self->pwm->SHORTS = NRF_PWM_SHORT_LOOPSDONE_SEQSTART0_MASK;
+    self->pwm->TASKS_SEQSTART[0] = 1;
+    self->playing = true;
+    self->stopping = false;
+    self->paused = false;
+}
+
+void common_hal_audiopwmio_pwmaudioout_stop(audiopwmio_pwmaudioout_obj_t* self) {
+    deactivate_audiopwmout_obj(self);
+    self->pwm->TASKS_STOP = 1;
+    self->stopping = false;
+    self->paused = false;
+
+    m_free(self->buffers[0]);
+    self->buffers[0] = NULL;
+
+    m_free(self->buffers[1]);
+    self->buffers[1] = NULL;
+}
+
+bool common_hal_audiopwmio_pwmaudioout_get_playing(audiopwmio_pwmaudioout_obj_t* self) {
+    if(!self->paused && self->pwm->EVENTS_STOPPED) {
+        self->playing = false;
+        self->pwm->EVENTS_STOPPED = 0;
+    }
+    return self->playing;
+}
+
+/* pause/resume present difficulties for the NRF PWM audio module.
+ *
+ * A PWM sequence can be stopped in its tracks by sending a TASKS_STOP event,
+ * but there's no way to pick up the sequence where it was stopped; you could
+ * start at the start of one of the two sequences, but especially for "single buffer"
+ * sample, this seems undesirable.
+ *
+ * Or, you can stop at the end of a sequence so that you don't duplicate anything
+ * when restarting, but again this is unsatisfactory for a "single buffer" sample.
+ *
+ * For now, I've taken the coward's way and left these methods unimplemented.
```

Comment:
You've implemented them! Is this accurate now? One change for looped samples would be that it's desirable to stop on the boundary of the sample.

Reply:
Comment updated to reflect the current state of the code.